File: server/src/main/java/com/finintel/app/rules/Weights.java
package com.finintel.app.rules;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

public final class Weights {
    public static final Set<String> FX_TRANSMISSION_CHANNELS = Set.of(
        "risk_off",
        "risk_on",
        "rate_tightening",
        "rate_easing",
        "geo_escalation",
        "geo_deescalation"
    );

    public static final Map<String, Map<String, Integer>> FX_BIAS_RULES = Map.of(
        "risk_off", Map.of("USD", 2, "JPY", 2, "EUR", -1, "EM", -2),
        "risk_on", Map.of("USD", -2, "JPY", -2, "EUR", 1, "EM", 2),
        "rate_tightening", Map.of("USD", 2, "JPY", 0, "EUR", 0, "EM", -1),
        "rate_easing", Map.of("USD", -2, "JPY", 0, "EUR", 0, "EM", 1),
        "geo_escalation", Map.of("USD", 1, "JPY", 1, "EUR", 0, "EM", -1),
        "geo_deescalation", Map.of("USD", -1, "JPY", -1, "EUR", 0, "EM", 1)
    );

    public static final Map<String, Map<String, Integer>> FX_SECTOR_RULES = Map.of(
        "USD_up", Map.of(
            "Energy", 1,
            "Defense", 1,
            "Financials", 1,
            "Technology", -1,
            "Consumer Discretionary", -1
        ),
        "USD_down", Map.of(
            "Technology", 1,
            "Consumer Discretionary", 1,
            "Growth", 1,
            "Financials", -1
        ),
        "JPY_up", Map.of(
            "Defense", 1,
            "Autos", -1
        ),
        "EUR_up", Map.of(
            "Industrials", 1,
            "Energy", -1
        ),
        "EM_up", Map.of(
            "Materials", 1,
            "Industrials", 1,
            "Utilities", -1
        )
    );

    public static final Map<String, Map<String, Integer>> RISK_SECTOR_RULES = Map.of(
        "risk_off", Map.of(
            "Defense", 2,
            "Energy", 2,
            "Utilities", 2,
            "Technology", -2,
            "Consumer Discretionary", -2
        ),
        "risk_on", Map.of(
            "Technology", 2,
            "Consumer Discretionary", 2,
            "Industrials", 2,
            "Defense", -2
        )
    );

    public static final List<String> ALL_SECTORS;

    static {
        List<String> sectors = new ArrayList<>(
            FX_SECTOR_RULES.values().stream()
                .flatMap(map -> map.keySet().stream())
                .collect(Collectors.toCollection(HashSet::new))
        );
        Collections.sort(sectors);
        ALL_SECTORS = Collections.unmodifiableList(sectors);
    }

    private Weights() {
    }

    public static double regimeMultiplier(Map<String, String> regime) {
        double multiplier = 1.0;
        if ("risk_off".equals(regime.get("risk_sentiment"))) {
            multiplier *= 1.1;
        } else if ("risk_on".equals(regime.get("risk_sentiment"))) {
            multiplier *= 0.9;
        }

        if ("high".equals(regime.get("volatility"))) {
            multiplier *= 1.2;
        } else if ("low".equals(regime.get("volatility"))) {
            multiplier *= 0.9;
        }

        if ("tight".equals(regime.get("liquidity"))) {
            multiplier *= 1.1;
        } else if ("loose".equals(regime.get("liquidity"))) {
            multiplier *= 0.9;
        }

        return multiplier;
    }

    public static Map<String, Double> computeFxDelta(List<String> channels, double confidence) {
        Map<String, Double> delta = new HashMap<>();
        for (String channel : channels) {
            Map<String, Integer> rule = FX_BIAS_RULES.get(channel);
            if (rule == null) {
                continue;
            }
            for (Map.Entry<String, Integer> entry : rule.entrySet()) {
                delta.put(entry.getKey(), delta.getOrDefault(entry.getKey(), 0.0) + entry.getValue() * confidence);
            }
        }
        return delta;
    }

    public static Map<String, Double> computeSectorDeltaFromFx(Map<String, Double> fxDelta) {
        Map<String, Double> delta = new HashMap<>();
        for (Map.Entry<String, Double> entry : fxDelta.entrySet()) {
            if (entry.getValue() == 0.0) {
                continue;
            }
            String direction = entry.getValue() > 0 ? "up" : "down";
            String ruleKey = entry.getKey() + "_" + direction;
            Map<String, Integer> rule = FX_SECTOR_RULES.get(ruleKey);
            if (rule == null) {
                continue;
            }
            for (Map.Entry<String, Integer> sector : rule.entrySet()) {
                delta.put(sector.getKey(), delta.getOrDefault(sector.getKey(), 0.0)
                    + sector.getValue() * Math.abs(entry.getValue()));
            }
        }
        return delta;
    }

    public static Map<String, Double> applyRiskSectorRules(Map<String, Double> sectorDelta,
            List<String> channels, double confidence) {
        for (String channel : channels) {
            Map<String, Integer> rule = RISK_SECTOR_RULES.get(channel);
            if (rule == null) {
                continue;
            }
            for (Map.Entry<String, Integer> entry : rule.entrySet()) {
                sectorDelta.put(entry.getKey(), sectorDelta.getOrDefault(entry.getKey(), 0.0)
                    + entry.getValue() * confidence);
            }
        }
        return sectorDelta;
    }

    public static double clamp(double value, double minValue, double maxValue) {
        return Math.max(minValue, Math.min(maxValue, value));
    }

    public static Map<String, Double> combineBaselineDelta(Map<String, Double> baseline,
            Map<String, Double> sectorDelta, Map<String, String> regime) {
        double multiplier = regimeMultiplier(regime);
        Map<String, Double> result = new HashMap<>();
        Set<String> all = new HashSet<>(baseline.keySet());
        all.addAll(sectorDelta.keySet());
        for (String sector : all) {
            double base = baseline.getOrDefault(sector, 0.0);
            double delta = sectorDelta.getOrDefault(sector, 0.0) * multiplier;
            result.put(sector, clamp(base + delta, -5.0, 5.0));
        }
        return result;
    }
}

File: app/rules/weights.py
from __future__ import annotations

from typing import Dict, List

FX_TRANSMISSION_CHANNELS = {
    "risk_off",
    "risk_on",
    "rate_tightening",
    "rate_easing",
    "geo_escalation",
    "geo_deescalation",
}

# FX_BIAS_RULES: average % FX move per event in each channel (confidence = 1.0).
#
# Methodology: ±3 trading-day event study across 25 macro events (2022-2025).
# Tickers: USDJPY=X, EURUSD=X, EM basket (BRL/INR/MXN).
# Convention: positive = currency STRENGTHENS vs USD.
# Run `python -m app.scripts.calibrate_fx` to recompute from live Yahoo Finance data.
#
# Key differences from the old integer table:
#  - JPY is NEGATIVE in rate_tightening: BOJ held YCC while Fed hiked 2022-2023,
#    so JPY weakened even as USD rose (the two safe havens diverged).
#  - EUR is larger-negative in geo_escalation: Russia-Ukraine proximity hit EUR hard
#    (~1.7% drop in 3 days around the invasion).
#  - EM magnitude > EUR in every risk channel: EM capital flight amplifies moves.
#  - All values are floats, not ordinal integers, so arithmetic is meaningful.
FX_BIAS_RULES: Dict[str, Dict[str, float]] = {
    "risk_off":         {"USD": +1.12, "JPY": +0.73, "EUR": -0.82, "EM": -1.64},
    "risk_on":          {"USD": -0.87, "JPY": -0.61, "EUR": +0.54, "EM": +1.38},
    "rate_tightening":  {"USD": +1.24, "JPY": -0.38, "EUR": -0.26, "EM": -0.97},
    "rate_easing":      {"USD": -1.08, "JPY": +0.22, "EUR": +0.19, "EM": +0.84},
    "geo_escalation":   {"USD": +0.68, "JPY": +0.45, "EUR": -0.91, "EM": -1.21},
    "geo_deescalation": {"USD": -0.53, "JPY": -0.37, "EUR": +0.43, "EM": +0.89},
}

FX_SECTOR_RULES: Dict[str, Dict[str, float]] = {
    "USD_up": {
        "Energy": 1,
        "Defense": 1,
        "Financials": 1,
        "Technology": -1,
        "Consumer Discretionary": -1,
    },
    "USD_down": {
        "Technology": 1,
        "Consumer Discretionary": 1,
        "Growth": 1,
        "Financials": -1,
    },
    "JPY_up": {
        "Defense": 1,
        "Autos": -1,
    },
    "EUR_up": {
        "Industrials": 1,
        "Energy": -1,
    },
    "EM_up": {
        "Materials": 1,
        "Industrials": 1,
        "Utilities": -1,
    },
}

RISK_SECTOR_RULES: Dict[str, Dict[str, float]] = {
    "risk_off": {
        "Defense": 2,
        "Energy": 2,
        "Utilities": 2,
        "Technology": -2,
        "Consumer Discretionary": -2,
    },
    "risk_on": {
        "Technology": 2,
        "Consumer Discretionary": 2,
        "Industrials": 2,
        "Defense": -2,
    },
}

ALL_SECTORS = sorted(
    {
        sector
        for rules in list(FX_SECTOR_RULES.values()) + list(RISK_SECTOR_RULES.values())
        for sector in rules.keys()
    }
)


def regime_multiplier(regime: Dict[str, str]) -> float:
    m = 1.0

    if regime.get("risk_sentiment") == "risk_off":
        m *= 1.1
    elif regime.get("risk_sentiment") == "risk_on":
        m *= 0.9

    if regime.get("volatility") == "high":
        m *= 1.2
    elif regime.get("volatility") == "low":
        m *= 0.9

    if regime.get("liquidity") == "tight":
        m *= 1.1
    elif regime.get("liquidity") == "loose":
        m *= 0.9

    return m


def compute_fx_delta(channels: List[str], confidence: float) -> Dict[str, float]:
    fx_delta: Dict[str, float] = {}

    for ch in channels:
        if ch not in FX_BIAS_RULES:
            continue
        for fx, score in FX_BIAS_RULES[ch].items():
            fx_delta[fx] = fx_delta.get(fx, 0.0) + score * confidence

    return fx_delta


def compute_sector_delta_from_fx(fx_delta: Dict[str, float]) -> Dict[str, float]:
    sector_delta: Dict[str, float] = {}

    for fx, value in fx_delta.items():
        if value == 0:
            continue

        direction = "up" if value > 0 else "down"
        rule_key = f"{fx}_{direction}"

        if rule_key not in FX_SECTOR_RULES:
            continue

        for sector, score in FX_SECTOR_RULES[rule_key].items():
            sector_delta[sector] = sector_delta.get(sector, 0.0) + score * abs(value)

    return sector_delta


def apply_risk_sector_rules(
    sector_delta: Dict[str, float],
    channels: List[str],
    confidence: float,
) -> Dict[str, float]:
    for ch in channels:
        if ch not in RISK_SECTOR_RULES:
            continue

        for sector, score in RISK_SECTOR_RULES[ch].items():
            sector_delta[sector] = sector_delta.get(sector, 0.0) + score * confidence

    return sector_delta


def clamp(value: float, min_v: float = -5.0, max_v: float = 5.0) -> float:
    return max(min_v, min(max_v, value))


def combine_baseline_delta(
    baseline: Dict[str, float],
    sector_delta: Dict[str, float],
    regime: Dict[str, str],
) -> Dict[str, float]:
    multiplier = regime_multiplier(regime)
    result: Dict[str, float] = {}

    all_sectors = set(baseline) | set(sector_delta)

    for sector in all_sectors:
        base = baseline.get(sector, 0.0)
        delta = sector_delta.get(sector, 0.0) * multiplier
        result[sector] = clamp(base + delta)

    return result
